tracer/src/Datadog.Trace/Debugger/RateLimiting/AdaptiveSampler.cs

```diff
@@ -0,0 +1,261 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Threading;
+
+namespace Datadog.Trace.Debugger.RateLimiting
+{
+    /// <summary>
+    /// An adaptive streaming (non-remembering) sampler.
+    ///
+    /// The sampler attempts to generate at most N samples per fixed time window in randomized
+    /// fashion. For this it divides the timeline into 'sampling windows' of constant duration. Each
+    /// sampling window targets a constant number of samples which are scattered randomly (uniform
+    /// distribution) throughout the window duration and once the window is over the real stats of
+    /// incoming events and the number of gathered samples is used to recompute the target probability to
+    /// use in the following window.
+    ///
+    /// This will guarantee, if the windows are not excessively large, that the sampler will be able
+    /// to adjust to the changes in the rate of incoming events.
+    ///
+    /// However, there might so rapid changes in incoming events rate that we will optimistically use
+    /// all allowed samples well before the current window has elapsed or, on the other end of the
+    /// spectrum, there will be to few incoming events and the sampler will not be able to generate the
+    /// target number of samples.
+    ///
+    /// To smooth out these hiccups the sampler maintains an under-sampling budget which can be used
+    /// to compensate for too rapid changes in the incoming events rate and maintain the target average
+    /// number of samples per window.
+    /// </summary>
+    internal class AdaptiveSampler
+    {
+        /// <summary>
+        /// Exponential Moving Average (EMA) last element weight.
+        /// Check out papers about using EMA for streaming data - eg.
+        /// https://nestedsoftware.com/2018/04/04/exponential-moving-average-on-streaming-data-4hhl.24876.html
+        ///
+        /// Corresponds to 'lookback' of N values:
+        /// With T being the index of the most recent value the lookback of N values means that for all values with index
+        /// T-K, where K > N, the relative weight of that value computed as (1 - alpha)^K is less or equal than the
+        /// weight assigned by a plain arithmetic average (= 1/N).
+        /// </summary>
+        private readonly double _emaAlpha;
+        private readonly int _samplesPerWindow;
+
+        private Counts _countsRef;
+
+        private double _probability = 1;
+
+        private long _samplesBudget;
+
+        private double _totalCountRunningAverage;
+        private double _avgSamples;
+
+        private int _budgetLookback;
+        private double _budgetAlpha;
+
+        private int _countsSlotIndex = 0;
+        private Counts[] _countsSlots;
+
+        private Timer _timer;
+        private Action _rollWindowCallback;
+
+        private ThreadLocal<Random> _rng;
```

Comment:
Great question! `Random` is not thread safe. See also https://stackoverflow.com/a/11109361

Reply:
One thing about this @OmerRaviv

IIUC `ThreadLocal<Random>` initializes _all_ the `Random` instances at once. As `Random` uses a time-based seed, that means that multiple threads could actually be using the same seed, and be generating the same "random" sequence?

I needed a thread-safe Random recently so used the approach here: https://github.com/DataDog/dd-trace-dotnet/blob/4ea25867575980af5d9e77c9f4a2a88cbfd3c79d/tracer/src/Datadog.Trace/Util/ThreadSafeRandom.cs

This uses the "built-in" thread-safe Random in .NET 6+, and uses a "global" `Random` to initialize the "threadlocal" `Random`s.

Reply:
Thank you! good to know about `ThreadSafeRandom`, looks neat 👍 

Yep, that's true! I'm not super concerned about multiple `AdaptiveSampler` generating the same random sequence, because all we are currently using the `Random`s for is to make an effort to make sampling decisions for debugger probes at different moments in time within the 1 second window interval (that should be more or less evenly distributed).  So this wouldn't matter much for our purpose.

Having said that, it's always good to have just one solution to a problem across the code base, and we _might_ end up reusing the AdaptiveSampler in other places where this would matter, so I'll submit a quick PR to reuse `ThreadSafeRandom` when I get back.

Reply:
Cool, I wasn't sure how much it mattered 🙂  I haven't actually merged that ThreadSafeRandom yet, so no rush, I can always do it 🙂